src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,569 @@
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the generic RPC Response listener could not be
+    /// registered with the given transport.
+    pub async fn new(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
+    ) -> Result<Self, UStatus> {
+        let store = InMemoryPendingRequestStore {
+            pending_requests: Mutex::new(HashMap::new()),
+        };
+        Self::with_pending_request_store(transport, uri_provider, Arc::new(store)).await
+    }
+
+    async fn with_pending_request_store(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
+        pending_request_store: Arc<dyn PendingRequestStore>,
+    ) -> Result<Self, UStatus> {
+        let response_listener = Arc::new(ResponseListener {
+            pending_request_store,
+        });
+        transport
+            .register_listener(
+                &UUri::any(),
+                Some(&uri_provider.get_source_uri()),
+                response_listener.clone(),
+            )
+            .await?;
+
+        Ok(InMemoryRpcClient {
+            transport,
+            uri_provider,
+            response_listener,
+        })
+    }
+}
+
+#[async_trait]
+impl RpcClient for InMemoryRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
+        call_options: CallOptions,
+        payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        let message_id = call_options.message_id().unwrap_or_else(UUID::build);
+
+        let mut builder = UMessageBuilder::request(
+            method.clone(),
+            self.uri_provider.get_source_uri(),
+            call_options.ttl(),
+        );
+        builder.with_message_id(message_id.clone());
+        if let Some(token) = call_options.token() {
+            builder.with_token(token.to_owned());
+        }
+        if let Some(priority) = call_options.priority() {
+            builder.with_priority(priority);
+        }
+        if let Some(traceparent) = call_options.traceparent() {
+            builder.with_traceparent(traceparent);
+        }
+        let rpc_request_message = if let Some(pl) = payload {
+            let format = pl.payload_format();
+            builder.build_with_payload(pl.payload(), format)
+        } else {
+            builder.build()
+        }
+        .map_err(|e| ServiceInvocationError::InvalidArgument(e.to_string()))?;
+
+        let response_future = self
+            .response_listener
+            .try_add_pending_request(message_id.clone())?;
+        self.transport
+            .send(rpc_request_message)
+            .await
+            .map_err(|e| {
+                self.response_listener.remove_pending_request(&message_id);
+                e
+            })?;
+        tokio::time::timeout(
```

Comment:
Would it be easier if we put the response data into async queue according to its reqid in the listener, and then retrieve the data here?

Reply:
Not sure I can follow. Can you elaborate a little?

Reply:
I'm not sure whether I missed anything, but I'm considering the possibility of avoiding implementing the Future. Here are my thoughts.

```rust
async fn invoke_method(...) {
     // Get (tx, rx) from async_channel
     // Put (reqid, tx) into HashMap
     // Waiting rx with tokio timeout
}

async fn on_receive(...) {
    // Use reqid to get tx in the HashMap
    // Put UMessage into tx
}
```

Reply:
Sounds like a great idea, let me try if I get this running ...

Reply:
I got this up and running, @evshary. I will update the PR soon ...

Reply:
would you mind taking another look @evshary ?